doc/gvpectrl.8
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` 
.    ds C' 
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GVPECTRL 8"
.TH GVPECTRL 8 "2008-09-01" "2.2" "GNU Virtual Private Ethernet"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
\&\f(CW\*(C`gvpectrl\*(C'\fR \- \s-1GNU\s0 Virtual Private Ethernet Control Program
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\f(CW\*(C`gvpectrl\*(C'\fR [\fB\-ckgs\fR] [\fB\-\-config=\fR\fI\s-1DIR\s0\fR] [\fB\-\-generate\-keys\fR] [\fB\-\-help\fR]
[\fB\-\-kill\fR[\fB=\fR\fI\s-1SIGNAL\s0\fR]] [\fB\-\-show\-config\fR] [\fB\-\-version\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This is the control program for the \f(CW\*(C`gvpe\*(C'\fR, the virtual private ethernet daemon.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-c\fR, \fB\-\-config=\fR\fI\s-1DIR\s0\fR" 4
.IX Item "-c, --config=DIR"
Read configuration options from \fI\s-1DIR\s0\fR.
.IP "\fB\-g\fR, \fB\-\-generate\-keys\fR" 4
.IX Item "-g, --generate-keys"
Generate public/private \s-1RSA\s0 key-pair and exit.
.IP "\fB\-q\fR, \fB\-\-quiet\fR" 4
.IX Item "-q, --quiet"
Suppresses messages the author finds nonessential for scripting purposes.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
Display short list of options.
.IP "\fB\-\-kill\fR[\fB=\fR\fI\s-1SIGNAL\s0\fR]" 4
.IX Item "--kill[=SIGNAL]"
Attempt to kill a running \f(CW\*(C`gvpectrl\*(C'\fR (optionally with the specified
\&\fI\s-1SIGNAL\s0\fR instead of \f(CW\*(C`SIGTERM\*(C'\fR) and exit.
.IP "\fB\-\-show\-config\fR" 4
.IX Item "--show-config"
Show a summary of the configuration, and how gvpe interprets it. Can also be very useful
when designing firewall scripts.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Output version information and exit.
.SH "BUGS"
.IX Header "BUGS"
If you find any bugs, report them to \f(CW\*(C`gvpe@schmorp.de\*(C'\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIgvpe\fR\|(5), \fIgvpe.conf\fR\|(5), \fIgvpe\fR\|(8).
.PP
\&\s-1GVPE\s0 comes with \s-1ABSOLUTELY\s0 \s-1NO\s0 \s-1WARRANTY\s0.  This is free software, and you are
welcome to redistribute it under certain conditions; see the file \s-1COPYING\s0
for details.
.SH "AUTHOR"
.IX Header "AUTHOR"
Marc Lehmann \f(CW\*(C`<gvpe@schmorp.de>\*(C'\fR.


NEWS
GVPE NEWS

2.21 Wed Sep  3 06:56:27 CEST 2008
	- add missing ev++.h include header to tarball, which everybody
          who tested it apparently had in their include path :(. Caught
          by Karl Kleinpaste.

2.2  Mon Sep  1 06:28:09 CEST 2008
	- protocol version 0.1, compatible with older releases,
          but upgrade is recommended to due changed ondemand behaviour.
        - new per-node options max-ttl and max-queue.
        - convert from iom.C to libev, a high-performance event loop
          (http://software.schmorp.de/pkg/libev).
        - tcp connections were leaking in some cases.
        - retry more aggressively (once/s) to establish a connection if
          new packets arrive for it.
        - save a lot of setsockopt calls when the tos doesn't change.
        - honor disabled even on initial connect attempt.
        - changed callback mechanism to be slightly less portable
          but more efficient mechanism (standards-compliant c++ compilers
          should work).
        - increased receive window positive size, to allow for massive
          packet loss due to occasional longer drop-outs.
        - send RST when a positive window size violation is detected, but
          not in other cases, to reconnect more quickly.
        - upgraded liblzf to version 3.4.
        - dropped -fno-exceptions due to ev++.h using it.
        - node-up/down scripts are now run in sequence.
        - new -q switch for gvpectrl, for when you run it often.
        - work around the horribly inconsistent, ad-hoc, ever-changing
          and broken texinfo syntax. YMMV. avoid texinfo.
        - keepalive is more aggressive now, sensding ping's every 3 seconds
          and killing the conenction after 15 seconds.
        - bugfixes.

2.01 Thu Mar 29 19:26:04 CEST 2007
	- protocol version 0.1, compatible with older releases.
        - bugfix of callback.h, might have cause callback return values to
          be corrupted on architectures like sparc before.
        - dns transport retries more aggressively.
        - updated documentation, improved dns transport reliability
          and throughput.
        - added experimental support for sha256 and sha512 digests.

2.0 Mon Dec  5 13:59:26 CET 2005
	- protocol version 0.1, compatible with older releases.
        - implement allow-direct, deny-direct node config statements.
        - implemented != for sockinfo. This fixes a bug where
          gvpe sent packets to the old ip address of another host
          even though it had received packets from it's new address.
          This only causes problems if you forget to -HUP your gvpe after
          your ip address changed, which is *required*.
        - sets close-on-exec flag on tcp connections. This fixes a bug
          where child processes kept tcp connections open and caused
          connections to fail when only one side can connect.
        - fixed a bug in receive sequence checking that made gvpe
          accept out-of-window packets in most cases.
        - tighter limit for the maximum sequence # to avoid overflow
          conditions + allow more headroom for packet reordering.
        - replace some asserts that trapped config mismatches by
          more useful log messages.
        - fix spurious extra newline in some log messages.

1.9 Tue Apr 19 06:21:50 CEST 2005
	- protocol version 0.1, compatible with older releases.
        - WARNING: this version checks the return value of if-up etc.
          scripts and exits on failure.
        - IMPORTANT: run if-up/node-up etc. scripts via /bin/sh.
        - IMPORTANT: interface initialization (MAC, MTU) is now done
          automatically in most configurations.
        - options can now be specified on the gvpe command line, too.
        - make some DNS transport values configurable and document them.
        - improved OS specific information in gvpe.osdep(5).
        - new tap device type "native/darwin", that supports the tap
          driver available for darwin (thanks to matthew mead who tested
          it out with me in a long session). tincd/darwin is still available.
        - new device type "tincd/bsd", which is a newer version of the
          *bsd-drivers taken from tinc.
        - fixed a bug in relying on the order of global construction
          when tcp transport was enabled. The fix makes it use no
          cpu time unless it is in use, too.
        - information about other nodes is now available to if-up etc. scripts.
        - the value of the config variable if-up-data is passed to the if-up
          etc. scripts.
	- skip unparsable config lines with a warning instead of stopping
          parsing and continuing with a certainly unusable config.

1.8 Fri Mar 18 00:58:55 CET 2005
	- protocol version 0.1, compatible with older releases.
	- enable-udp = yes is now default only when no other protocols
          are enabled. otherwise it is disabled unless explicitly enabled.
        - implemented dns tunneling (experimental now and in the future).
        - remove support for pre-release version protocol.
        - updated tincd drivers (rev 1433), added uml_socket driver, documented
          tincd drivers a bit better.
        - document icmp configuration values.
        - document transport protocols in gvpe.protocol(7).
        - remove unused ChangeLog file.
        - created a mailinglist at gvpe@lists.schmorp.de.
        - added an exemption to allow distribution of binaries linked against
          OpenSSL, as suggested by Guus Sliepen (author of tinc). No
          GNUTLS conversion in sight.
        - some portability fixes with respect to --disable-nls.

1.7 Tue Feb 22 23:58:59 CET 2005
	- protocol version 0.1, compatible with older releases.
	- first gnu release.
        - documented the special value 1 for router-priority.
        - renamed vped => gvpe and vpectrl => gvpectrl, as well as
          vped.conf => gvpe.conf.
	- new per-node option "max-retry".

src/gvpectrl.C
/*
    gvpectrl.C -- the main file for gvpectrl
    Copyright (C) 1998-2002 Ivo Timmermans <ivo@o2w.nl>
                  2000-2002 Guus Sliepen <guus@sliepen.eu.org>
                  2003-2008 Marc Lehmann <gvpe@schmorp.de>
 
    This file is part of GVPE.

    GVPE is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 3 of the License, or (at your
    option) any later version.
   
    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
    Public License for more details.
   
    You should have received a copy of the GNU General Public License along
    with this program; if not, see <http://www.gnu.org/licenses/>.
   
    Additional permission under GNU GPL version 3 section 7
   
    If you modify this Program, or any covered work, by linking or
    combining it with the OpenSSL project's OpenSSL library (or a modified
    version of that library), containing parts covered by the terms of the
    OpenSSL or SSLeay licenses, the licensors of this Program grant you
    additional permission to convey the resulting work.  Corresponding
    Source for a non-source form of such a combination shall include the
    source code for the parts of OpenSSL used as well as that of the
    covered work.
*/

#include "config.h"

#include <cstdio>
#include <cstring>
#include <cstdlib>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>

#include <openssl/rand.h>
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/evp.h>

#include "pidfile.h"

#include "conf.h"
#include "slog.h"
#include "util.h"
#include "vpn.h"

/* If nonzero, display usage information and exit. */
static int show_help;

/* If nonzero, print the version on standard output and exit.  */
static int show_version;

/* If nonzero, it will attempt to kill a running gvpe and exit. */
static int kill_gvpe;

/* If nonzero, it will attempt to kill a running gvpe and exit. */
static int show_config;

/* If nonzero, do not output anything but warnings/errors/very unusual conditions */
static int quiet;

/* If nonzero, generate public/private keypair for this net. */
static int generate_keys;

static struct option const long_options[] =
    {
      {"config", required_argument, NULL, 'c'},
      {"kill", optional_argument, NULL, 'k'},
      {"help", no_argument, &show_help, 1},
      {"version", no_argument, &show_version, 1},
      {"generate-keys", no_argument, NULL, 'g'},
      {"quiet", no_argument, &quiet, 1},
      {"show-config", no_argument, &show_config, 's'},
      {NULL, 0, NULL, 0}
    };

static void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help\' for more information.\n"), get_identity ());
  else
    {
      printf (_("Usage: %s [option]...\n\n"), get_identity ());
      printf (_
              ("  -c, --config=DIR           Read configuration options from DIR.\n"
               "  -k, --kill[=SIGNAL]        Attempt to kill a running gvpe and exit.\n"
               "  -g, --generate-keys        Generate public/private RSA keypair.\n"
               "  -s, --show-config          Display the configuration information.\n"
               "  -q, --quiet                Be quite quiet.\n"
               "      --help                 Display this help and exit.\n"
               "      --version              Output version information and exit.\n\n"));
      printf (_("Report bugs to <gvpe@schmorp.de>.\n"));
    }

  exit (status);
}

void
parse_options (int argc, char **argv, char **envp)
{
  int r;
  int option_index = 0;

  while ((r = getopt_long (argc, argv, "c:k::qgs", long_options, &option_index)) != EOF)
    {
      switch (r)
        {
        case 0:		/* long option */
          break;

        case 'c':		/* config file */
          confbase = strdup (optarg);
          break;

        case 'k':		/* kill old gvpes */
          if (optarg)
            {
              if (!strcasecmp (optarg, "HUP"))
                kill_gvpe = SIGHUP;
              else if (!strcasecmp (optarg, "TERM"))
                kill_gvpe = SIGTERM;
              else if (!strcasecmp (optarg, "KILL"))
                kill_gvpe = SIGKILL;
              else if (!strcasecmp (optarg, "USR1"))
                kill_gvpe = SIGUSR1;
              else if (!strcasecmp (optarg, "USR2"))
                kill_gvpe = SIGUSR2;
              else if (!strcasecmp (optarg, "INT"))
                kill_gvpe = SIGINT;
              else if (!strcasecmp (optarg, "ALRM"))
                kill_gvpe = SIGALRM;
              else
                {
                  kill_gvpe = atoi (optarg);

                  if (!kill_gvpe)
                    {
                      fprintf (stderr,
                               _
                               ("Invalid argument `%s'; SIGNAL must be a number or one of HUP, TERM, KILL, USR1, USR2, WINCH, INT or ALRM.\n"),
                               optarg);
                      usage (1);
                    }
                }
            }
          else
            kill_gvpe = SIGTERM;

          break;

        case 'g':		/* generate public/private keypair */
          generate_keys = RSA_KEYBITS;
          break;

        case 's':
          show_config = 1;
          break;

        case 'q':
          quiet = 1;
          break;

        case '?':
          usage (1);

        default:
          break;
        }
    }
}

/* This function prettyprints the key generation process */

void
indicator (int a, int b, void *p)
{
  if (quiet)
    return;

  switch (a)
    {
    case 0:
      fprintf (stderr, ".");
      break;

    case 1:
      fprintf (stderr, "+");
      break;

    case 2:
      fprintf (stderr, "-");
      break;

    case 3:
      switch (b)
        {
        case 0:
          fprintf (stderr, " p\n");
          break;

        case 1:
          fprintf (stderr, " q\n");
          break;

        default:
          fprintf (stderr, "?");
        }
      break;

    default:
      fprintf (stderr, "?");
    }
}

/*
 * generate public/private RSA keypairs for all hosts that don't have one.
 */
int
keygen (int bits)
{
  RSA *rsa_key;
  FILE *f;
  char *name = NULL;
  char *fname;

  asprintf (&fname, "%s/hostkeys", confbase);
  mkdir (fname, 0700);
  free (fname);

  asprintf (&fname, "%s/pubkey", confbase);
  mkdir (fname, 0700);
  free (fname);

  for (configuration::node_vector::iterator i = conf.nodes.begin (); i != conf.nodes.end (); ++i)
    {
      conf_node *node = *i;

      asprintf (&fname, "%s/pubkey/%s", confbase, node->nodename);

      f = fopen (fname, "a");

      if (!f)
        {
          perror (fname);
          exit (EXIT_FAILURE);
        }

      if (ftell (f))
        {
          if (!quiet)
            fprintf (stderr, "'%s' already exists, skipping this node %d\n",
                     fname, quiet);

          fclose (f);
          continue;
        }

      fprintf (stderr, _("generating %d bits key for %s:\n"), bits,
               node->nodename);

      rsa_key = RSA_generate_key (bits, 0xFFFF, indicator, NULL);

      if (!rsa_key)
        {
          fprintf (stderr, _("error during key generation!\n"));
          return -1;
        }
      else
        fprintf (stderr, _("Done.\n"));

      require (PEM_write_RSAPublicKey (f, rsa_key));
      fclose (f);
      free (fname);

      asprintf (&fname, "%s/hostkeys/%s", confbase, node->nodename);

      f = fopen (fname, "a");
      if (!f)
        {
          perror (fname);
          exit (EXIT_FAILURE);
        }

      require (PEM_write_RSAPrivateKey (f, rsa_key, NULL, NULL, 0, NULL, NULL));
      fclose (f);
      free (fname);
    }

  return 0;
}

int
main (int argc, char **argv, char **envp)
{
  set_identity (argv[0]);
  log_to (LOGTO_STDERR);

  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  parse_options (argc, argv, envp);

  if (show_version)
    {
      printf (_("%s version %s (built %s %s, protocol version %d.%d)\n"), get_identity (),
              VERSION, __DATE__, __TIME__, PROTOCOL_MAJOR, PROTOCOL_MINOR);
      printf (_("Built with kernel interface %s/%s.\n"), IFTYPE, IFSUBTYPE);
      printf (_
              ("Copyright (C) 2003 Marc Lehmann <gvpe@schmorp.de> and others.\n"
               "See the AUTHORS file for a complete list.\n\n"
               "vpe comes with ABSOLUTELY NO WARRANTY.  This is free software,\n"
               "and you are welcome to redistribute it under certain conditions;\n"
               "see the file COPYING for details.\n"));

      return 0;
    }

  if (show_help)
    usage (0);

  {
    configuration_parser (conf, false, 0, 0);
  }

  if (generate_keys)
    {
      RAND_load_file ("/dev/urandom", 1024);
      exit (keygen (generate_keys));
    }

  if (kill_gvpe)
    exit (kill_other (kill_gvpe));

  if (show_config)
    {
      conf.print ();
      exit (EXIT_SUCCESS);
    }

  usage (1);
}
